### PR TITLE
Generic class for multi key support

### PR DESCRIPTION
**Pull request by xiangyan99**
The change in Metrics Advisor code is not part of the multi-key support change.

I want to use it to demo how SDK developers will use the new feature.

### CODE REVIEW

File: sdk/core/azure-core/azure/core/credentials.py [hunk 1]
```diff
@@ -54,41 +53,40 @@ def get_token(
...
+T = TypeVar("T", str, Tuple[str, ...])
+
 
-class AzureKeyCredential:
+class AzureKeyCredential(Generic[T]):
     """Credential type used for authenticating to an Azure service.
     It provides the ability to update the key without creating a new client.
 
-    :param str key: The key used to authenticate to an Azure service
+    :param T key: The key used to authenticate to an Azure service
```
Comment (iscai-msft) [inline sdk/core/azure-core/azure/core/credentials.py:63]: is this the correct type for sphinx?

File: sdk/core/azure-core/azure/core/credentials.py [hunk 2]
```diff
@@ -54,41 +53,40 @@ def get_token(
...
+    def update(self, key: T) -> None:
         """Update the key.
-
         This can be used when you've regenerated your service key and want
         to update long-lived clients.
-
-        :param str key: The key used to authenticate to an Azure service
-        :raises: ValueError or TypeError
+        :param T key: The key used to authenticate to an Azure service
+        :raises: ValueError
```
Comment (iscai-msft) [inline sdk/core/azure-core/azure/core/credentials.py:84]: still raise a type error

File: sdk/core/azure-core/azure/core/credentials.py [hunk 3]
```diff
@@ -54,41 +53,40 @@ def get_token(
...
         """
         if not key:
             raise ValueError("The key used for updating can not be None or empty")
-        if not isinstance(key, str):
-            raise TypeError("The key used for updating must be a string.")
+        if not isinstance(key, (str, tuple)):
```
Comment (iscai-msft) [inline sdk/core/azure-core/azure/core/credentials.py:88]: can it be a tuple of anything? or does it have to be a tuple of strings

File: sdk/core/azure-core/azure/core/credentials.py [hunk 4]
```diff
@@ -54,41 +53,40 @@ def get_token(
...
+    :param T key: The key used to authenticate to an Azure service
     :raises: TypeError
     """
 
-    def __init__(self, key: str) -> None:
-        if not isinstance(key, str):
-            raise TypeError("key must be a string.")
-        self._key = key
+    def __init__(self, key: T) -> None:
+        if not isinstance(key, (str, tuple)):
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/credentials.py:68]: Sorry, still find the tuple solution ugly.

What about *args, like we do for scope:
```python
def __init__(self, key, *keys)
```

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:68]: Then what is the update story?

For now, we require uses to update all of them as an atomic operation.

if we have def update(self, key, *keys), does that mean we accept arbitrary keys?

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:68]: And to make sure we are on the same page, if it is a one-key scenario, user can directly use AzureKeyCredential with 

cred = AzureKeyCredential(key=<str>)

For multi-key scenarios, user does not use AzureKeyCredential directly, instead, SDK developers should have a custom class which is a subclass of AzureKeyCredential and expose the custom credential to the users.

Could you give more details about in which case, the user experience is ugly?

File: sdk/metricsadvisor/azure-ai-metricsadvisor/azure/ai/metricsadvisor/_patch.py [hunk 5]
```diff
@@ -27,7 +28,7 @@
 _X_API_KEY_HEADER_NAME = "x-api-key"
 
 
-class MetricsAdvisorKeyCredential:
+class MetricsAdvisorKeyCredential(AzureKeyCredential):
```
Comment (xiangyan99) [inline sdk/metricsadvisor/azure-ai-metricsadvisor/azure/ai/metricsadvisor/_patch.py:31]:
```suggestion
class MetricsAdvisorKeyCredential(AzureKeyCredential[(str,str)]:
```

**Pull request closed without merging**